node/src/runner.rs

```diff
@@ -105,17 +98,87 @@ pub async fn run(options: Options) -> Result<()> {
-                .context("Failed to initialize config")?;
-            log::trace!("Starting with the following config\n: {node_config:#?}");
-            let manager = initialize(run_options.net, node_config).await?;
-            #[allow(clippy::unit_arg)]
-            Ok(manager.main().await)
+        Command::Mainnet(ref run_options) => {
+            let chain_config = common::chain::config::create_mainnet();
+            start(&options.config_path(), run_options, chain_config).await
+        }
+        Command::Regtest(ref regtest_options) => {
+            let chain_config = regtest_chain_config(&regtest_options.chain_config)?;
+            start(
+                &options.config_path(),
+                &regtest_options.run_options,
+                chain_config,
+            )
+            .await
         }
     }
 }
 
+async fn start(
+    config_path: &Path,
+    run_options: &RunOptions,
+    chain_config: ChainConfig,
+) -> Result<()> {
+    let node_config =
+        NodeConfig::read(config_path, run_options).context("Failed to initialize config")?;
+    log::trace!("Starting with the following config\n: {node_config:#?}");
+    let manager = initialize(chain_config, node_config).await?;
+    manager.main().await;
+    Ok(())
+}
+
+fn regtest_chain_config(options: &ChainConfigOptions) -> Result<ChainConfig> {
+    let ChainConfigOptions {
+        chain_address_prefix,
+        chain_blockreward_maturity,
+        chain_max_future_block_time_offset,
+        chain_version,
+        chain_target_block_spacing,
+        chain_coin_decimals,
+        chain_max_block_header_size,
+        chain_max_block_size_with_standard_txs,
+        chain_max_block_size_with_smart_contracts,
+    } = options;
+
+    let mut builder = ChainConfigBuilder::new(ChainType::Regtest);
+
+    macro_rules! update_builder {
+        ($field: ident) => {
+            paste! {
+                if let Some(val) = [<chain_ $field>] {
+                    builder = builder.$field(val.to_owned());
+                }
+            }
+        };
+        ($field: ident, $converter: stmt) => {
+            paste! {
+                if let Some(val) = [<chain_ $field>] {
+                    builder = builder.$field($converter(val.to_owned()));
+                }
+            }
+        };
+        ($field: ident, $converter: stmt, map_err) => {
+            paste! {
+                if let Some(val) = [<chain_ $field>] {
+                    builder = builder.$field($converter(val.to_owned()).map_err(|e| anyhow!(e))?);
+                }
+            }
+        };
+    }
+
+    update_builder!(address_prefix);
+    update_builder!(blockreward_maturity, BlockDistance::new);
+    update_builder!(max_future_block_time_offset, Duration::from_secs);
+    update_builder!(version, SemVer::try_from, map_err);
+    update_builder!(target_block_spacing, Duration::from_secs);
+    update_builder!(coin_decimals);
```

Comment:
I am not sure configuring `coin_decimals` will work correctly. It should be treated as a read-only constant or removed form chain_config for the time of being.

Reply:
Even though coin_decimals is "too constant" to be treated like a variable, but I don't see an issue there. The only issue I see happening is with the reward schedule. What else could happen?

Reply:
I have just removed this option, but I can bring it back based on the result of this discussion. 🙃

Reply:
It does not really have any effect in reward schedule, so I think it might be a rather misleading setting with poorly defined effects.

Reply:
Also it seems to me to be more of a UI concern rather than something node should be concerned with.

Reply:
Remember that we want to test everything. Later, with rpc, we want to test how the wallet works. All this has to be tested together. It's not an option for normal ChainConfig. This is just for tests.

I'm OK with removing this option now and being able to bring it later if we have to. But I prefer to have the option than not. Having it and not using it is better than needing it and not having it.